jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,275 @@
+            }
+        except KeyError:
+            self.log.info("No route for {} doesn't exist!".format(routespec))
+            result = None
+        finally:
+            return result
+
+    def _persist_routes(self):
+        with traefik_utils.atomic_writing(self.toml_dynamic_config_file) as config_fd:
+            config_fd.write("[frontends]\n")
+            for key, value in self.routes_cache.items():
+                config_fd.write("".join(value["frontend"]))
+            config_fd.write("[backends]\n")
+            for key, value in self.routes_cache.items():
+                config_fd.write("".join(value["backend"]))
+
+    def _update_config_file(self):
+        try:
+            tmp = open("tmp.toml", "w")
+            self._persist_routes(tmp)
+            tmp.flush()
+            os.fsync(tmp.fileno())
+            tmp.close()
+            os.replace("tmp.toml", self.toml_dynamic_config_file)
+        except:
+            self.log.error("Failed to update traefik's dynamic configuration file")
+
+    async def start(self):
+        """Start the proxy.
+
+        Will be called during startup if should_start is True.
+
+        **Subclasses must define this method**
+        if the proxy is to be started by the Hub
+        """
+        await super().start()
+        await self._wait_for_static_config(provider="file")
+
+    async def stop(self):
+        """Stop the proxy.
+
+        Will be called during teardown if should_start is True.
+
+        **Subclasses must define this method**
+        if the proxy is to be started by the Hub
+        """
+        await super().stop()
+        self._clean_resources()
+
+    async def add_route(self, routespec, target, data):
+        """Add a route to the proxy.
+
+        **Subclasses must define this method**
+
+        Args:
+            routespec (str): A URL prefix ([host]/path/) for which this route will be matched,
+                e.g. host.name/path/
+            target (str): A full URL that will be the target of this route.
+            data (dict): A JSONable dict that will be associated with this route, and will
+                be returned when retrieving information about this route.
+
+        Will raise an appropriate Exception (FIXME: find what?) if the route could
+        not be added.
+
+        The proxy implementation should also have a way to associate the fact that a
+        route came from JupyterHub.
+        """
+
+        self.log.info("Adding route for %s to %s.", routespec, target)
+
+        (
+            backend_alias,
+            backend_url_path,
+            frontend_alias,
+            frontend_rule_path,
+        ) = traefik_utils.generate_route_keys(self, target, routespec, separator=".")
+
+        data = json.dumps(data)
+        rule = traefik_utils.generate_rule(routespec)
+
+        await self.mutex.acquire()
+        try:
+            self.routes_cache[routespec] = {
+                "backend": [
```

Comment:
Instead of toml-as-text, let's use a dict here, and `toml.dump` to persist it. We may even want to use `toml.load` to read it in from disk to start.